src/block-data-providers/bitcoin-core/provider.ts

```diff
@@ -0,0 +1,237 @@
+            await this.setState(state);
+        }
+    }
+
+    @Cron(CronExpression.EVERY_10_SECONDS)
+    async sync() {
+        if (this.isSyncing) return;
+        this.isSyncing = true;
+
+        const state = await this.getState();
+        if (!state) {
+            throw new Error('State not found');
+        }
+        const tipHeight = await this.getTipHeight();
+        if (tipHeight <= state.indexedBlockHeight) {
+            this.logger.debug(
+                `No new blocks found. Current tip height: ${tipHeight}`,
+            );
+            this.isSyncing = false;
+            return;
+        }
+
+        let height = state.indexedBlockHeight + 1;
+
+        for (height; height <= tipHeight; height++) {
+            const transactions = await this.processBlock(height);
+            for (const transaction of transactions) {
+                const { txid, vin, vout, blockHeight, blockHash } = transaction;
+                await this.indexTransaction(
+                    txid,
+                    vin,
+                    vout,
+                    blockHeight,
+                    blockHash,
+                );
+            }
+            state.indexedBlockHeight = height;
+            await this.setState(state);
+        }
+        this.isSyncing = false;
+    }
+
+    private async getTipHeight(): Promise<number> {
+        const body = {
+            method: 'getblockcount',
+            params: [],
+        };
+        return this.request(body);
+    }
+
+    private async getBlockHash(height: number): Promise<string> {
+        const body = {
+            method: 'getblockhash',
+            params: [height],
+        };
+        return this.request(body);
+    }
+
+    private async getBlock(hash: string, verbosity: number): Promise<Block> {
+        const body = {
+            method: 'getblock',
+            params: [hash, verbosity],
+        };
+        return this.request(body);
+    }
+
+    private async getRawTransaction(
+        txid: string,
+        isVerbose: boolean,
+    ): Promise<BlockTransaction> {
+        const body = {
+            method: 'getrawtransaction',
+            params: [txid, isVerbose],
+        };
+        return this.request(body);
+    }
+
+    public async processBlock(height: number): Promise<Transaction[]> {
+        const parsedTransactionList: Transaction[] = [];
+        const blockHash = await this.getBlockHash(height);
+        this.logger.debug(
+            `Processing block at height ${height}, hash ${blockHash}`,
+        );
+        const block = await this.getBlock(blockHash, 2);
+        const excludedCoinbaseTxns = block.tx.slice(1);
+        for (const txn of excludedCoinbaseTxns) {
+            const parsedTransaction = await this.parseTransaction(
+                txn,
+                block.hash,
+                block.height,
+            );
+            parsedTransactionList.push(parsedTransaction);
+        }
+        return parsedTransactionList;
```

Comment:
I see that here you are creating an array of transactions and returning to the caller. 
Then in the caller, you are iterating over the array and indexing the transactions.
I feel that you should do the indexing right here.

Reply:
This is done that way to ensure consistency, that is  when error happens in processing a block, the database state doesn't remain inconsistent by having some transactions in a block processed

Reply:
yes, I get it but in future, we want the ability to be able to resume processing from a partially processed block. This would require storing the index of the last processed tx in the block.

Reply:
personally i would have liked if the index could be done in a blockwise fashion, so there can never be an inconsistent state. With my current impl, what could make an incosistent state is  this `await this.transactionsService.saveTransaction(transaction);`. If saving all blocks could be done atomically in the DB, there would never be an inconsistent state

Reply:
I hear you. IMHO:
- indexing a full block is a costly operation right now because we are making a lot of network requests
- the chances of failure is high because there are a lot of network requests

that's why I don't feel that indexing a block should be an atomic operation

Reply:
But in your followup PR, if use verbosity 3, then there will only be one network call, and the chances of failure will be low.
So when we make that change, the constraint that block indexing should be atomic will make sense